File: ASSIGNMNET6/python/Q.1.py
class Stack:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stack = []
    
    def is_empty(self):
        """Check if the stack is empty."""
        return len(self.stack) == 0
    
    def is_full(self):
        """Check if the stack is full."""
        return len(self.stack) == self.capacity
    
    def push(self, item):
        """Add an item to the top of the stack."""
        if self.is_full():
            print("Stack Overflow! Cannot push", item)
        else:
            self.stack.append(item)
            print(f"Pushed {item} onto stack.")
    
    def pop(self):
        """Remove and return the top item of the stack."""
        if self.is_empty():
            print("Stack Underflow! Cannot pop from an empty stack.")
            return None
        else:
            item = self.stack.pop()
            print(f"Popped {item} from stack.")
            return item
    
    def peek(self):
        """Return the top item of the stack without removing it."""
        if self.is_empty():
            print("Stack is empty! Cannot peek.")
            return None
        else:
            return self.stack[-1]
    
    def display(self):
        """Display the contents of the stack."""
        if self.is_empty():
            print("Stack is empty!")
        else:
            print("Stack contents:", self.stack)

# Example usage
if __name__ == "__main__":
    stack_capacity = 5
    stack = Stack(stack_capacity)

    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.display()

    print("Top element is:", stack.peek())
    
    stack.pop()
    stack.display()
    
    stack.pop()
    stack.pop()
    stack.pop()  # This should show underflow message

    stack.push(40)
    stack.display()

File: ASSIGNMNET6/c/Q.1.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 5  // Define the maximum size of the stack

// Stack structure definition
typedef struct Stack {
    int items[MAX];
    int top;
} Stack;

// Function to initialize the stack
void initializeStack(Stack *stack) {
    stack->top = -1;  // Stack is initially empty
}

// Check if the stack is empty
int isEmpty(Stack *stack) {
    return stack->top == -1;
}

// Check if the stack is full
int isFull(Stack *stack) {
    return stack->top == MAX - 1;
}

// Function to push an element onto the stack
void push(Stack *stack, int value) {
    if (isFull(stack)) {
        printf("Stack is full! Cannot push %d\n", value);
    } else {
        stack->items[++stack->top] = value;
        printf("Pushed %d onto the stack.\n", value);
    }
}

// Function to pop an element from the stack
int pop(Stack *stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty! Cannot pop.\n");
        return -1;  // Return -1 to indicate the stack is empty
    } else {
        return stack->items[stack->top--];
    }
}

// Function to peek at the top element of the stack
int peek(Stack *stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty! Nothing to peek.\n");
        return -1;  // Return -1 to indicate the stack is empty
    } else {
        return stack->items[stack->top];
    }
}

// Main function to test the stack implementation
int main() {
    Stack stack;
    initializeStack(&stack);

    // Push elements onto the stack
    push(&stack, 10);
    push(&stack, 20);
    push(&stack, 30);
    push(&stack, 40);
    push(&stack, 50);
    push(&stack, 60);  // This push should fail as the stack is full

    // Peek at the top element
    printf("Top element is: %d\n", peek(&stack));

    // Pop elements from the stack
    printf("Popped element: %d\n", pop(&stack));
    printf("Popped element: %d\n", pop(&stack));
    printf("Popped element: %d\n", pop(&stack));

    // Peek at the top element again
    printf("Top element after popping is: %d\n", peek(&stack));

    return 0;
}
